utils.py

```diff
@@ -0,0 +1,487 @@
+import csv
+import glob
+import multiprocessing as mp
+import os
+from typing import Tuple
+import warnings
+
+from absl import logging
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import scipy.io
+import tqdm    
+import xlmhg
+
+import istarmap
+
+
+warnings.simplefilter(action='ignore', category=FutureWarning)
+
+def check_data_type(data_path: str) -> str:
+    """Checks if data is single cell or spatial.
+    
+    If 'filtered_feature_bc_matrix' exists in the expected folder, 
+    data type is considered spatial, and scRNAseq otherwise.
+
+    Args:
+        data_path: path for dataset folder.
+
+    Returns:
+        the data type of the provided dataset, "scRNAseq" or "spatial".
+    """
+
+    counts_path = os.path.join(data_path, "filtered_feature_bc_matrix")
+    is_spatial = os.path.exists(counts_path)
+    if is_spatial:
+        return 'spatial'
+    else:
+        return 'scRNAseq'
+
+
+def detect_data_file(data_path: str) -> str:
+    """Checks scRNAseq file extention and returns detected file name.
+    
+    Supported extensions: txt, tsv or pkl.
+    Prioratizing for pkl file, assuming it contains processed data.
+
+    Args:
+        data_path: path for dataset.
+
+    Return:
+        the detected file name.
+
+    Raises:    
+        Exception if more than one file was detected.
```

Comment:
use (existing or custom) less broad exceptions. 
if only one type is raised, keep it in a single declaration: BlaBlaError if this happened or if that happened.